tools/bpf2c/bpf_code_generator.cpp

```diff
@@ -1494,6 +1660,19 @@ bpf_code_generator::emit_c_code(std::ostream& output_stream)
             output_stream << std::endl;
         }
 
+        if (programs.size() > program_count) {
+            output_stream << "// Forward references for local functions." << std::endl;
+            for (auto& [_, subprogram] : programs) {
+                if (is_subprogram(subprogram)) {
+                    output_stream << "static uint64_t" << std::endl;
+                    output_stream << subprogram.program_name.c_identifier()
+                                  << "(uint64_t r1, uint64_t r2, uint64_t r3, uint64_t r4, uint64_t r5, uint64_t r10);"
```

Comment:
This should be (r10 - stack_used) by current function. This is why local variables get overwritten.

Reply:
Discussed offline.
Blocked on https://github.com/vbpf/ebpf-verifier/issues/720 but we think this PR is a useful subset to merge in the meantime.